Fesom2.6 recom tra diags and CI test

ogurses opened the pull request.

ogurses committed:
feat(fesom) add tracer diagnostics to FESOM2.5

Define new variables to track tracer changes
due to advection and diffusion.

We want to save for now diffusion and advection
contribution to the tracer changes. Horizontal and
vertical diffusion includes Redi
parametrization (if it is set .true.).

ogurses committed:
add recom_sms diag

ogurses committed:
feat(recom): add ciso and medusa components

Fill __ciso directive to ensure that
carbon isotope code works. Medusa interface is
added.

a270105 committed:
input to bc_surface from medusa added and Sed_input deleted

ogurses committed:
fix(recom) Add missing directives

ogurses committed:
two small fixes

ogurses committed:
feat(recom)!: FESOM2.6-REcoM setup with 4p3z2d

Use __coccos in src/CMakeLists.txt to
include 2 more phytoplankton classes

FESOM2.6-REcoM setup with 4 phytoplankton,
3 zooplankton and 2 detritus functional types.

BREAKING CHANGE: FESOM2.6-REcoM setup with 4p3z2d

ogurses committed:
fix(recom) Add missing temperature function code

Correct temperature function for small phytoplankton and diatoms.

Previous update for the fouth phytoplankton class and
new temperature functions for small phytoplankton and diatoms
resulted in very low Phaeocystis biomass.
We recognised that this is related to missing code lines in the
temperature function which we correct with this update.

ogurses committed:
fix(recom): Add detritus sinking velocity and K0, solubility to diagnostics

Use namelist.io to activate diagnostics.

Sinking velocity of particle (fast and slow) are calculated but not written into a file before.
We fix the  missing diagnostics for sinking velocity of detritus classes. Besides we
Extracted the K0, solubility  which is computed
in gasx.F90.

ogurses committed:
Merge remote-tracking branch 'origin/fesom2.6_recom_tra_diags' into fesom2.6_recom_tra_diags

ogurses committed:
fix(recom) clean merge conflicts

ogurses committed:
fix(recom): Correct silicate assimilation and diatom diagnostics 

We add __coccos directive to get the correct temperature
function in silicate assimilation.

In the coccos and Phaeocystis version, it should be dependent on the new temperature function, Temp_diatoms.
Use the original Arrhenius function in simpler phytoplankton version. Besides, light limitation diagnostics is corrected. It does affect only the diatoms related output.

ogurses committed:
fix(recom): Correct Light limitation diagnostics(continued)

ogurses committed:
fix(recom): Correct namelist.recom files in config folder

ogurses committed:
fix(fesom): Bugfix: sign of real salt flux

Correct reported salt flux bug in oce_ale_tracer.F90.
https://github.com/FESOM/fesom2/issues/721

ogurses committed:
fix(recom):  Return to saturation model for chlorophyll degradation

We commented out the linear model where degradation is directly
proportional to light intensity and the chlorophyll degradation
rate is capped at a maximum of 0.3. Reason for this is the unrealistic
air2sea co2 fluxes in the 2p3z2d setup.
The reactivated saturation model is more realistic. It accounts for light
saturation effects.
At low light, degradation increases roughly linearly with PAR
At high light, the degradation rate approaches an asymptotic maximum

ogurses committed:
feat(recom)!: feat(recom)!: replace `__coccos` and `__3Zoo2Det` directives with namelist parameters

Replace preprocessor directives with namelist parameters:
* `__3Zoo2Det` → `enable_3zoo2det`
* `__coccos` → `enable_coccos`

* Enhanced inline comments with ecological significance and biological context
*Complete variable/parameter documentation with units and descriptions
* Clear sectioning with visual separators and variable definitions
* Equation explanations showing mathematical relationships and biological processes
* Logic documentation for decision trees and conditional blocks
* Comments now explain "why" not just "what"

**Sequence Validation**
* Generates expected tracer ID list based on configuration
* Shows exact ID order and position-by-position comparison

**Configuration-Specific Clash Detection**
* Full Model (both flags): Don't use IDs 1023-1024
* Coccos-only: Use 1023-1028 (NOT 1029-1034)
* 3Zoo2Det-only: Use 1029-1030 for microzoo (NOT 1035-1036)

**Validation Checks**
* Exact sequence matching against expected IDs
* Duplicate ID detection
*Forbidden ID detection per configuration

BREAKING CHANGE: replace  directives with namelist parameter for model configuration

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tra_diags

JanStreffing committed:
bit more git like

JanStreffing committed:
Remove comment from files array declaration

ogurses committed:
ci(recom): Add initialization and surface ocean boundary conditions for REcoM-CI test

Add recom initialzation and sbc files for recom. Standard setup for recom is

enable_3zoo2det = .true.  and enable_coccos = .false. (namelist.recom)

suvarchal committed:
Merge branch 'main' into fesom2.6_recom_tra_diags

suvarchal committed:
Merge branch 'main' into fesom2.6_recom_tra_diags

suvarchal commented:
@ogurses can you also provide a namelist.tra that uses files in test/input/recom i see in config/bin_2p1z1d/namelist.tra uses
```
&tracer_init3d                            ! initial conditions for tracers
n_ic3d   = 8                              ! number of tracers to initialize
idlist   = 1019, 1022, 1018, 1003, 1002, 1001, 2, 1 ! their IDs (0 is temperature, 1 is salinity, etc.). The reading order is defined here!
filelist = 'fe_pisces_opa_eq_init_3D_changed_name.nc', 'woa18_all_o00_01_mmol_fesom2.nc', 'woa13_all_i00_01_fesom2.nc', 'GLODAPv2.2016b.TAlk_fesom2_mmol_fix_z_Fillvalue.nc', 'GLODAPv2.2016b.TCO2_fesom2_mmol_fix_z_Fillvalue.nc', 'woa13_all_n00_01_fesom2.nc', 'phc3.0_winter.nc', 'phc3.0_winter.nc'  ! list of files in ClimateDataPath to read (one file per tracer), same order as idlist
varlist  = 'Fe', 'oxygen_mmol', 'i_an', 'TAlk_mmol', 'TCO2_mmol', 'n_an', 'salt', 'temp'                 ! variables to read from specified files
t_insitu = .true.                         ! if T is insitu it will be converted to potential after reading it
/
```
but fe_pieces*.nc is not in test/input.

JanStreffing committed:
wip building recom ci test

JanStreffing committed:
Fix REcoM CI test: add push trigger and improve debugging

JanStreffing committed:
Add REcoM namelist templates to fix CI test

- namelist.recom and namelist.tra required by REcoM (not auto-generated)
- Updated file paths to use local copies in work directory
- Set REcoM_restart=false for initialization from climatology
- Workflow copies these template files to work_recom/

JanStreffing committed:
Use default REcoM namelists from config/ instead of duplicates

- Copy namelists from config/bin_2p3z2d/ (the actual defaults)
- Patch file paths with sed for CI environment
- Remove duplicate namelist files from setups/test_pi_recom/
- This ensures we test with real default REcoM configuration

JanStreffing committed:
Use simpler 2p1z1d REcoM configuration for faster CI test

- Switch from bin_2p3z2d (30 tracers) to bin_2p1z1d (22 tracers)
- Reduces computation time by ~30% while still testing core REcoM
- bgc_num: 30 -> 22 (2 phyto, 1 zoo, 1 detritus instead of 3 zoo, 2 det)
- Faster CI runtime for basic validation

JanStreffing committed:
Fix sed patterns to avoid breaking namelist format

- Use more precise patterns that match exact variable assignments
- Clear unused file paths (aen, river, erosion) we don't have
- Escape dots properly for .true./.false. matching
- Add debug output to show patched namelist
- Previous greedy patterns were breaking pavariables section

JanStreffing committed:
Fix namelist.tra to use correct test input files

- namelist.tra was referencing non-existent files like fe_pisces_opa_eq_init_3D_changed_name.nc
- Patch filelist in tracer_init3d section to use our test files:
  fe5deg.nc, oxy5deg.nc, si5deg.nc, talk5deg.nc, tco2_5deg.nc, din5deg.nc, woa18_netcdf_5deg.nc
- Add debug output to show patched tracer initialization

JanStreffing committed:
Simplify namelist patching with cleaner sed patterns

- Use simpler wildcard patterns to match file paths
- Use sed 'c\' command to replace entire filelist line in namelist.tra
- Remove excessive debug output, keep only essential verification
- Should avoid any line length or parsing issues

JanStreffing committed:
Add namelist validation checks to debug pavariables error

JanStreffing committed:
Use pre-configured namelists instead of runtime patching

- Add namelist.recom.working and namelist.tra.working to test/input/recom/
- Based on branch owner's working versions with paths fixed for CI
- bgc_num=22 for 2p1z1d configuration
- REcoM_restart=false for initialization from climatology
- Avoids any runtime sed issues that might cause parsing errors

JanStreffing committed:
Debug: Check if mkrun creates namelist.recom or if we need to add it

JanStreffing committed:
Fix: Copy and patch namelists since mkrun doesn't create them

- mkrun does NOT create namelist.recom or namelist.tra for REcoM
- Copy from config/bin_2p1z1d/ and patch all file paths
- Replace individual filenames in namelist.tra with our test files
- Fix REcoM_restart=false and REcoMDataPath='./'
- This should finally work\!

JanStreffing committed:
Fix REcoM namelist reading bug

Line 178 was incorrectly reading 'pavariables' namelist when it should
read 'parecomsetup', causing pavariables to be read twice. This resulted
in 'Invalid line in namelist' and EOF errors when trying to run REcoM.

Fixed by correcting the namelist read on line 178 from pavariables to
parecomsetup, so the reading order matches the namelist file structure.

JanStreffing committed:
Simplify CI workflow: remove unused nam_rsbc section

The code doesn't read nam_rsbc (no variables defined), so remove it
from the namelist to avoid parsing errors. This also simplifies the
workflow - we only need to patch REcoM_restart and tracer files now.

JanStreffing committed:
Switch to 2p3z2d (30 tracers) configuration for REcoM CI test

The 2p1z1d (22 tracers) configuration causes segfaults during initialization,
likely due to arrays (wFluxPhy, wFluxDia, etc.) not being properly allocated
for the simplified configuration.

Use the full 2p3z2d (30 tracers) configuration which is better tested and
has all arrays properly allocated.

JanStreffing committed:
Fix REcoM segfault: allocate missing wFlux arrays

The wFlux* arrays (wFluxDet, wFluxPhy, wFluxDia, wFluxCocco, wFluxPhaeo)
and their global counterparts (GlowFlux*) were being initialized to 0.0
but were never allocated, causing a segmentation fault at line 137.

Added allocate() statements for all missing arrays before initialization.
This bug affected all REcoM configurations (2p1z1d and 2p3z2d).

JanStreffing committed:
Set test year to 1948 to match available forcing data

- Changed initial year from 1958 to 1948 in clockinit
- Updated REcoM forcing year range to 1948
- Added copying of 1948 forcing files in CI workflow

The test forcing data files are from 1948, so we need to start
the simulation in 1948 rather than 1958.

JanStreffing committed:
Remove push trigger from REcoM workflow

Only trigger REcoM CI test on pull requests to main branch
or manual workflow_dispatch. This reduces CI load during
development.

JanStreffing committed:
Disable leap years for REcoM test to match forcing data

The CORE forcing data uses noleap calendar (365 days/year).
Set include_fleapyear=False to match the forcing data calendar.

JanStreffing committed:
Fix segfault: add missing argument to monthly_event call

The monthly_event subroutine requires two arguments (do_output, N),
but was called with only one at line 1872 in gen_surface_forcing.F90.
This caused undefined behavior when accessing the uninitialized N
parameter, leading to a segmentation fault during forcing updates.

Fixed by adding N=1 (check every month) as the second argument.

JanStreffing committed:
Fix dust file name: create symlink for default filename

The code has a hardcoded default nm_fe_data_file='DustClimMonthlyAlbani.nc'
but our test file is named 'DustClimMonthlyAlbani_pimesh.nc'.

Since we remove the nam_rsbc section (it's not recognized by the Fortran
namelist parser), the code uses the default filename. Create a symlink
so the file is accessible under both names.

JanStreffing committed:
Fix dust file path: create symlink in ClimateDataPath

The dust forcing file is read from ClimateDataPath (test/input/recom/),
not from the working directory. Create the symlink in the correct location.

JanStreffing committed:
Fix segfault: allocate missing RiverFe array

The RiverFe array was declared in recom_modules.F90 but never allocated,
causing a segmentation fault at line 2080 in gen_surface_forcing.F90
when trying to set RiverFe = 0.0d0.

Added allocation and initialization for RiverFe alongside the other
river nutrient arrays.

JanStreffing committed:
Enable atmospheric pressure forcing for REcoM CO2 flux

Added mslp (mean sea level pressure) forcing configuration to enable
CO2 flux calculations in REcoM. Without atmospheric pressure, the CO2
flux calculation fails with NaN values.

Changes:
- Added namelist.forcing section with l_mslp=True
- Configured to read from 'slp.' files with 'slp' variable
- Copy PHC2_salx.nc and runoff.nc to work directory in CI

JanStreffing committed:
Fix NetCDF files with missing _FillValue attributes

Add preprocessing step to process REcoM input NetCDF files using
'cdo copy' (with ncks as fallback) to add proper _FillValue attributes.

Files processed:
- oxy5deg.nc
- si5deg.nc
- talk5deg.nc
- din5deg.nc
- GLODAPv2.2016b.PI_TCO2_fesom2_mmol_fix_z_Fillvalue.nc

This ensures proper land masking and interpolation instead of using
dummy fill values (10000000000.0 and 0.0).

Suggested by colleague to fix interpolation issues.

JanStreffing committed:
Fix NetCDF files with proper _FillValue attributes

Processed REcoM input files with 'cdo copy' to add proper fill values:
- oxy5deg.nc
- si5deg.nc
- talk5deg.nc
- din5deg.nc
- GLODAPv2.2016b.PI_TCO2_fesom2_mmol_fix_z_Fillvalue.nc

This is a one-time fix to the repository files, removing the need
for runtime preprocessing in the CI workflow.

JanStreffing committed:
Enable atmospheric pressure forcing by patching namelist.forcing

The default namelist.forcing.CORE2 already has mslp (atmospheric
pressure) configured but disabled (l_mslp=.false.). Patch it to
l_mslp=.true. in the workflow to enable reading of slp.*.nc files.

This is needed for REcoM CO2 flux calculations which require
atmospheric pressure (Patm).

Removed unnecessary namelist.forcing section from setup.yml since
we just patch the default config file directly.

JanStreffing committed:
Fix sed regex: escape dots in .false. pattern

The sed command wasn't matching because dots are regex wildcards.
Escaped the dots in .false. to match the literal string.

Now l_mslp=.false. will correctly be replaced with l_mslp=.true.

JanStreffing committed:
Add namelist.forcing display to verify l_mslp patch

Show l_mslp status in the test run output to diagnose why
atmospheric pressure is still not being read (Patm=0.0).

JanStreffing committed:
Fix sed pattern to handle spaces around = in l_mslp

The namelist.forcing file has:
   l_mslp = .false.  (with spaces)

Previous pattern l_mslp=\.false\. didn't match.
New pattern l_mslp *= *\.false\. matches optional spaces.

This should finally enable atmospheric pressure forcing\!

JanStreffing committed:
Reduce debug output in CI workflow

Namelists are working correctly now, so removed verbose dumps.
Kept minimal verification that patches were applied:
- REcoM_restart setting
- l_mslp atmospheric pressure flag

This will make the CI logs much cleaner and easier to read.

JanStreffing committed:
Fix NetCDF fill value handling for CDO/NCO compatibility

Problem: Fortran NetCDF library and CDO/NCO tools use different
fill value conventions, causing NaN values in tracer initialization.

Solution: Make fill value detection more robust:
1. Try _FillValue attribute (Fortran NetCDF standard)
2. Fall back to missing_value attribute (used by CDO/NCO)
3. Use NetCDF default fill value (9.969e+36) if neither exists
4. Broadcast fill value to all MPI ranks

This fixes the 'REcoM_DIC: NaN' error during initialization by
properly handling land/missing values in input NetCDF files
created with different tools.

Fixes issue reported by colleague about incompatible fill value
conventions between different NetCDF tools.

JanStreffing committed:
Add NaN debugging to DIC calculation in REcoM

Added diagnostic output to track where DIC becomes NaN:
- Check state(idic) and sms(idic) before DIC calculation
- Check DIC after calculation
- Print node, depth, lon/lat location when NaN detected
- Stop execution immediately to get clean traceback

This will help identify whether:
1. The state already contains NaN (initialization issue)
2. The SMS contains NaN (biogeochemistry issue)
3. The addition creates NaN (numerical issue)

The error occurs at lon=-43.88, lat=14.55 consistently, so
this should pinpoint the exact cause.

JanStreffing committed:
Add NaN debugging to recom_forcing before CO2 flux

Added diagnostic checks at the point where DIC is read from state
array and converted for CO2 flux calculation.

Since the SMS debug didn't trigger, the NaN must appear between
SMS calculation and forcing. This will catch:
1. If state(one,idic) is already NaN when forcing starts
2. If the conversion to REcoM_DIC creates NaN

Will print node number, coordinates, and related state values
to identify the exact source of NaN.

ogurses committed:
fix(recom): Bugfixes related to ballasting

When ballasting is used, vertical sinking is adjusted
if the detritus carbon biomass exceeds a treshold value (0.001).
This results in large air-sea co2flux uptake. Weremove this constraint.

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tra_diags

JanStreffing committed:
overwrite Sweeney path for recom ci test

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tra_diags

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tra_diags

JanStreffing committed:
Refactor tracer diag loops

patrickscholz commented:
I talked a while a ago with Sergey about this, and we should try to encapsulate as much of RECOM as we can, To achieve that as much of RECOM lives within the recom_...F90 code files and is only called by function and subroutines (e.g boundary conditions, initialisation, ...). We have been a bit sloppy about this with all the iceberg and transient tracer stuff in the past, which made large part of the FESOM code especially the initialisation,  tracer and boundary condition part very crowded and messy and more difficult to maintain.

JanStreffing commented:
Perfect task for the AI. I had it search for more cases:

Similar repeated-structure cases in REcoM code (candidates for future refactor)

1) src/int_recom/recom_main.F90
   - Pattern: In the main node loop (do n=1,myDim_nod2D) there is a recurring
     "backup state -> do work -> compute diagnostic delta" idiom.
   - Example: ttf_rhs_bak is filled for all tracers with ltra_diag, and later
     tracers%work%tra_recom_sms(...) is computed as values - ttf_rhs_bak.
   - Why it’s similar: conceptually the same as the backup/delta calculation
     refactored in src/oce_ale_tracer.F90 (just organized differently).

2) src/int_recom/recom_sinking.F90
   - Pattern: long sequences of if(tracer ID == ...) blocks repeat nearly the same
     update structure for different element pools:
       - update Benthos(n,k)
       - optionally update SinkFlx_tr(n,k,tr_num) when use_MEDUSA
       - optionally update Benthos_tr(n,k,tr_num)
     repeated for N/C/Si/CaCO3 and also isotope variants.
   - Why it’s similar: repetitive boilerplate where only the tracer-ID set and the
     target index (k) changes.

3) src/int_recom/recom_init.F90
   - Pattern: repeated min/max scanning over nodes for multiple REcoM tracers,
     followed by repeated MPI_Allreduce calls with identical structure.
   - Why it’s similar: repeated blocks of the same operation on different tracer IDs.

4) src/int_recom/recom_sms.F90 (benthic remineralization section)
   - Pattern: repeated update equations with the same shape:
       decayBenthos(i) = rate(i) * LocBenthos(i)
       LocBenthos(i)   = LocBenthos(i) - decayBenthos(i) * dt_b
     repeated across elements and (optionally) isotopes.
   - Why it’s similar: identical structure differing only by indices/parameters.

JanStreffing committed:
REcoM: deduplicate ltra_diag backup/delta

JanStreffing committed:
REcoM: deduplicate benthos accumulation

JanStreffing committed:
REcoM: deduplicate init sanity-check reductions

JanStreffing committed:
REcoM: deduplicate benthos decay updates

JanStreffing committed:
Merge pull request #863 from FESOM/recom_refactoring_dedup

Recom refactoring dedup

JanStreffing committed:
Add script to run REcoM CI test locally via Singularity

Runs the same steps as the GitHub Actions fesom2_recom workflow
inside the CI container on Levante. Requires pre-pulling the SIF
image on the login node (compute nodes have read-only TMPDIR).

JanStreffing committed:
Fix DIC NaN caused by uninitialized ErosionTOC2D array

Root cause: copy-paste bug in recom_init.F90 where ErosionTON2D was
initialized twice and ErosionTOC2D was never initialized. Uninitialized
memory (NaN) propagated through bc_surface for DIC via IEEE 754
NaN * 0.0 = NaN, even when erosion input was disabled.

Also adds a guard for nodes with nzmax < 1 in the REcoM main loop and
hardens tracer initialization in gen_ic3d.F90 to clean up NaN values
from CDO/NCO interpolation artifacts.

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tra_diags

JanStreffing committed:
move ci test run script to work folder

JanStreffing committed:
update truth

JanStreffing committed:
Fix REcoM CI output: use otracers for DIN/DIC and set correct reference values

The io_list used explicit "DIN" and "DIC" variable names, but the model
only outputs these via the generic "otracers" keyword. Changed to
"otracers" and updated fcheck reference values from actual CI run output.

JanStreffing committed:
Fix coupled_yac CI build: install libfyaml sub-headers

Upstream libfyaml (commit 902a981, March 3 2026) split the monolithic
libfyaml.h into an umbrella header that #includes 13 sub-headers from
libfyaml/. The install command only ran install-includeHEADERS (the
umbrella), not install-libfyamlHEADERS (the sub-headers). When YAC's
configure tested #include <libfyaml.h>, the compiler could not find the
sub-headers and reported "libfyaml.h not found".
